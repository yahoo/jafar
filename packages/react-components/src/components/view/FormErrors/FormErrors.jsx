/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import FormContext from '@jafar/react-form/context';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { withTheme } from '@material-ui/core/styles';

const defaultLabels = {
  required: 'Required',
  invalid: 'Invalid',
};

const Wrapper = styled.div`
  > div {
    padding-top: 20px;
    &:first-child {
      padding-top: 0px;
    }
  }
`;

const FieldTitle = styled.div`
  font-size: 12px;
  text-transform: uppercase;
`;

const FieldLink = withTheme(styled.div`
  list-style-type: none;
  margin-top: 10px;
  cursor: pointer;
  color: ${({ theme }) => theme.palette.secondary.main};
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  outline: none;
  &:hover {
    text-decoration: underline;
  };
`);


const Fields = ({
  fields, title, onClickField,
}) => (fields.length > 0 && <div>
  <FieldTitle>{title}</FieldTitle>
  { fields.map(field => (<FieldLink key={field.id} onClick={() => { onClickField(field.id); }}>{field.label}</FieldLink>)) }
</div>);

const mapFieldsErrors = (fields, fieldIds) => {
  const requiredFields = [];
  const invalidFields = [];

  Object.keys(fields)
    .filter(fieldId => !fieldIds || fieldIds.includes(fieldId))
    .forEach((fieldId) => {
      const field = fields[fieldId];
      const required = field.errors.find(err => err.name === 'required');
      if (required) {
        requiredFields.push({ id: fieldId, label: field.label, error: required });
      } else if (field.errors.length) {
        invalidFields.push({ id: fieldId, label: field.label, error: field.errors[0] });
      }
    });

  return { required: requiredFields, invalid: invalidFields };
};

/**
 * Showing Jafar form's errors
 * 
 * Import <a target="_blank" 
 href="https://github.com/yahoo/jafar/blob/master/packages/react-components/src/components/view/FormErrors/FormErrors.jsx">
 FormErrors</a> from '@jafar/react-components/view/FormErrors'
 */
function FormErrors({ labels = defaultLabels, onClickField, fields }) {
  const form = React.useContext(FormContext);
  const invalidFields = mapFieldsErrors(form.model.fields, fields);

  return (
    <Wrapper>
      <Fields fields={invalidFields.required} title={labels.required} onClickField={onClickField} isFirst={true} />
      <Fields fields={invalidFields.invalid} title={labels.invalid} onClickField={onClickField} />
    </Wrapper>
  );
}

FormErrors.propTypes = {
  labels: PropTypes.shape({
    required: PropTypes.string,
    invalid: PropTypes.string,
  }),
  onClickField: PropTypes.func,
  fields: PropTypes.arrayOf(PropTypes.string),
};

export default FormErrors;
