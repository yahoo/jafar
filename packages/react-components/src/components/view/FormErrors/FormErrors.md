<h3>Examples</h3>

<h4>Usage in jafar form</h4>

```javascript
const Text = require('../../edit/Text/index.js').default;
const Form = require('@jafar/react-form/Form').default;
const Field = require('@jafar/react-form/Field').default;

const model = {
  id: 'simple',
  fields: {
    firstName: {
      label: 'First Name',
      path: 'firstName',
      component: {
        name: 'Text',
      },
      required: true,
      validators: [{
        name: 'minLength',
        args: { value: 5 }
      }],
    },
  },
};

const resources = {
  components: { 
      Text: { renderer: Text }, 
  },
  terms: {
    excludeMe: {
      func: () => false, // convert to true to see the field excluded
    },
  },
};

<Form model={model} resources={resources}>
  <Field id='firstName' />
  <div style={{ marginTop: '40px' }}>
    <FormErrors onClickField={fieldId => console.log(fieldId)} />
  </div>
</Form>
```

