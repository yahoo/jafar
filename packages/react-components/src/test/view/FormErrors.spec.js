/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import { shallow } from 'enzyme';
import FormErrors from '../../components/view/FormErrors';

describe('<FormErrors />', () => {
  let component;
  let context;

  beforeEach(() => {
    FormErrors.contextTypes = {
      model: PropTypes.object,
    };
    context = {
      model: { 
        fields: {
          firstName: {
            label: 'First Name',
            errors: [{
              name: 'required',
              message: 'Field Required',
            }],
          },
          lastName: {
            label: 'Last Name',
            errors: [{
              name: 'required',
              message: 'Field Required',
            }],
          },
        },
        invalid: true,
      },
    };
    jest.spyOn(React, 'useContext').mockImplementation(() => context);
    
  });
  it('should render all invalid fields', () => {
    component = shallow(getComponent());
    expect(component).toMatchSnapshot();
  });

  it('should render some invalid fields', () => {
    component = shallow(getComponent(['firstName']));
    expect(component).toMatchSnapshot();
  });

  function getComponent(fields) {
    return (<FormErrors onClickField={() => {}} fields={fields} />);
  }
});
