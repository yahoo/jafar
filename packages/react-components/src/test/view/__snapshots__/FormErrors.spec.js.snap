// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormErrors /> should render all invalid fields 1`] = `
<styled.div>
  <Fields
    fields={
      Array [
        Object {
          "error": Object {
            "message": "Field Required",
            "name": "required",
          },
          "id": "firstName",
          "label": "First Name",
        },
        Object {
          "error": Object {
            "message": "Field Required",
            "name": "required",
          },
          "id": "lastName",
          "label": "Last Name",
        },
      ]
    }
    isFirst={true}
    onClickField={[Function]}
    title="Required"
  />
  <Fields
    fields={Array []}
    onClickField={[Function]}
    title="Invalid"
  />
</styled.div>
`;

exports[`<FormErrors /> should render some invalid fields 1`] = `
<styled.div>
  <Fields
    fields={
      Array [
        Object {
          "error": Object {
            "message": "Field Required",
            "name": "required",
          },
          "id": "firstName",
          "label": "First Name",
        },
      ]
    }
    isFirst={true}
    onClickField={[Function]}
    title="Required"
  />
  <Fields
    fields={Array []}
    onClickField={[Function]}
    title="Invalid"
  />
</styled.div>
`;
